### PR TITLE
feature/Create-Record-transform

### PR DESCRIPTION
**Pull request by ailegion**
added: Create Record transform

JIRA: https://issues.cask.co/browse/PLUGIN-391

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>io.cdap.plugin</groupId>
+  <artifactId>hierarchy-plugins</artifactId>
+  <version>1.0.0-SNAPSHOT</version>
+
+  <properties>
+    <!-- properties for script build step that creates the config files for the artifacts -->
+    <data.pipeline.parent>system:cdap-data-pipeline[6.1.1,7.0.0-SNAPSHOT)</data.pipeline.parent>
+    <data.stream.parent>system:cdap-data-streams[6.1.1,7.0.0-SNAPSHOT)</data.stream.parent>
```
Comment (rmstar) [inline pom.xml]: lower bound should match `cdap.version`

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+                                                  failureCollector);
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
+    }
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+    FailureCollector failureCollector = context.getFailureCollector();
+    createRecordTransformConfig.validate(failureCollector);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:151]: Add validation in `prepareRun()`. Also emit lineage.

File: docs/CreateRecord-transform.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"id": ["customer_id"],
+	"customer": {
+		"name": ["customer_name"],
+		"phone": ["customer_phone"]
+	},
+	"orders": {
+		"id": ["order_id"],
+		"product_id": ["product_id"],
+		"amount": ["amount"],
+		"description": ["order_description"]
```
Comment (rmstar) [inline docs/CreateRecord-transform.md:46]: Please provide an example mapping for the case where the hierarchy is more than 1 level deep. Please add a unit test for this case as well.

**Review by rmstar — COMMENTED** [review]
Please add screenshots of the plugin properties and documentation widgets.

File: pom.xml (hunk 1 above)
Comment (ailegion) [reply pom.xml]: fixed

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java (hunk 2 above)
Comment (ailegion) [reply src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:151]: added

File: docs/CreateRecord-transform.md (hunk 3 above)
Comment (ailegion) [reply docs/CreateRecord-transform.md:46]: updated

**Comment by ailegion** [comment]
> Please add screenshots of the plugin properties and documentation widgets.

Requested screenshots:
CreateRecords-Widgets
CreateTransform-DocumentationPart1
CreateTransform-DocumentationPart2

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java [hunk 4]
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
...
+
+  /**
+   * Generates lineage for each record in mapping json
+   *
+   * @param record
+   */
+  private LinkedList<FieldOperation> generateLineageRecord(String recordName, JsonElement record) {
+    LinkedList<FieldOperation> fieldOperations = new LinkedList<>();
+    List<String> inputFields = new ArrayList<>();
+    List<String> recordFields = new ArrayList<>();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: Is `recordFields` the list of output fields generated by the operation?

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java [hunk 5]
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
...
+    for (Map.Entry<String, JsonElement> recordNode : record.getAsJsonObject().entrySet()) {
+      JsonElement value = recordNode.getValue();
+      if (value.isJsonArray()) {
+        recordFields.add(recordNode.getKey());
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: please add a comment explaining what you're doing here

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java [hunk 6]
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
...
+        inputFields.add(value.getAsJsonArray().get(0).getAsString());
+      }
+      if (value.isJsonObject()) {
+        fieldOperations.addAll(generateLineageRecord(recordNode.getKey(), recordNode.getValue()));
+        recordFields.add(recordNode.getKey());
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: we update recordFields but not inputFields - is that correct? Please add a comment here explaining what we're doing here.

**Comment by CuriousVini** [comment]
I think recent merge caused conflicts. Could you please resolve conflicts too along with comments?

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java (hunk 4 above)
Comment (ailegion) [reply src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: updated - changed to ``outputFields``

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java (hunk 5 above)
Comment (ailegion) [reply src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: added comment

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java (hunk 6 above)
Comment (ailegion) [reply src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java]: added comment - when type is JSON object in map config that means we have nested field on output, we call the function recursively in order to map input fields and output fields.

**Comment by ailegion** [comment]
> I think recent merge caused conflicts. Could you please resolve conflicts too along with comments?

resolved the conflicts and addressed the comments.

File: src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+      JsonElement value = recordNode.getValue();
+      if (value.isJsonArray()) {
+        // the field is not nested field - input is mapped to output directly
+        outputFields.add(recordNode.getKey());
+        inputFields.add(value.getAsJsonArray().get(0).getAsString());
+      }
+      if (value.isJsonObject()) {
+        // the field is nested object - make recursive call to map nested object fields
+        fieldOperations.addAll(generateLineageRecord(recordNode.getKey(), recordNode.getValue()));
+        outputFields.add(recordNode.getKey());
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:437]: We are adding an output field here, but no input field. Is that correct?

For example, if we have input table:
```
| field                 | type         |
| ------------          | ------------ |
| customer_profile      | Record       |
| - customer_name       | String       |
| - customer_phone      | String       |
```

with mapping:
```	
"customer": {
		"name": ["customer_profile", "customer_name"],
		"phone": ["customer_profile", "customer_phone"]
}
```

we would add a FieldTransOperation with inputFields=[], outputFields=["customer"].

Comment (ailegion) [reply src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:437]: When it's nested record it calls recursive function (line 436) - which will hit condition on line 429 for fields inside the nested record and will get input and output fields and then create FieldTransformOperation in line 445 for the nested record. 

In the example you gave the "customer_profile" input field will be in the new FieldTransformOperation generated inside the recursive call.

**Review by rmstar — APPROVED** [review]
Please squash commits. Thanks!

**Comment by ailegion** [comment]
> Please squash commits. Thanks!

squashed

**Pull request merged**